borgmatic/config/schema.yaml

```diff
@@ -1265,3 +1265,28 @@ properties:
                     Umask used when executing hooks. Defaults to the umask that
                     borgmatic is run with.
                 example: 0077
+            snapper:
+                type: object
+                required: ['include']
+                additionalProperties: false
+                properties:
+                    include:
+                        anyOf:
+                            - type: array
+                            - type: string
+                              enum:
+                                - all
+                        description: | 
+                            List of source_directories to include or "all" to
+                            include all source_directories a snapper config
+                            exists for
+                    exclude:
+                        type: array
+                        description: | 
+                            List of source_directories to exclude. If a path is
+                            listed in include and exclude, exclude will take
+                            precedence
+                description: |
+                    Configuration for integration with snapper. All included
+                    directories will use their latest snapper snapshot for
+                    backup instead of the configured source directory
```

Comment:
Hmm, so this is where it gets ugly apparently :D
>Do you have thoughts on whether users will care that their snapshotted files will effectively be in a different location in the backup archive?

Yeah, this seems pretty annoying. One way to address this might be to call borg like this: `subprocess.run["borg", "create", repo, ".", cwd=source_directory` for each source directory. But I suppose this is completely different from how borgmatic does things currently.

With regards to the database dumps: Since the snapshots are taken before the dumps are created, I'm not sure how that would integrate with borgmatic at all. Excluding `~/.borgmatic` from snapshotting can not be done within borgmatic, but needs to be done in snapper config/btrfs subvolumes/mount hierarchy. This could be helped through documentation though I suppose.

But additionally `~/.borgmatic` would need to be readded to the source_directories such that the dumps will be included.

I need to think on this a bit but feel free to chime in with insights of your own if you have any.

Reply:
> Yeah, this seems pretty annoying.

I left an idea in another comment about bind mounts that might help address this.

> One way to address this might be to call borg like this: subprocess.run(["borg", "create", repo, "."], cwd=source_directory) for each source directory. But I suppose this is completely different from how borgmatic does things currently.

Interesting! Yeah, I think that would also result in the backed up path showing up differently in the archive. Because today IIRC the full `/source/directory/` is a prefix on every path in the archive.

> I need to think on [the database dump question] a bit but feel free to chime in with insights of your own if you have any.

Bind mount? :smiley: Then it doesn't matter if `~/.borgmatic` is snapshotted because the path won't change. (I admit I have a :hammer: and everything looks like a nail.)

Reply:
The bind mount idea is the best. It will also help in data deduplication and faster archive creation. However, will it be prudent to also store the absolute directory path in the comment (or in the metadata) of the archive? Because then we will be able to restore to the exact location.

Reply:
As much as I like the idea of bind mounts, I'm not sure that it would impact deduplication. My understanding is that Borg deduplicates at the [chunk level](https://borgbackup.readthedocs.io/en/stable/#main-features) regardless of filename.